File: scripts/build_index.py
#!/usr/bin/env python3
"""Build a compact knowledge index of the brain directory.

Scans People/, Projects/, Areas/, Notes/, Tasks/tasks.md, Journal/
and outputs formatted markdown tables for use as AI context.

Usage:
    python3 build_index.py              # Full index
    python3 build_index.py --compact    # Compact (names only, for hooks)
"""

import os
import re
import sys
from pathlib import Path

def _brain_path() -> Path:
    return Path(os.path.expanduser(os.environ.get("REMEMBER_BRAIN_PATH", "~/remember")))

def parse_frontmatter(filepath: Path, max_bytes=2048) -> dict:
    """Parse YAML frontmatter + H1 from first ~2KB of a file."""
    try:
        text = filepath.read_text(encoding="utf-8")[:max_bytes]
    except OSError:
        return {}
    
    meta = {}
    # Extract frontmatter
    if text.startswith("---"):
        end = text.find("---", 3)
        if end > 0:
            fm = text[3:end]
            for line in fm.strip().splitlines():
                if ":" in line:
                    key, _, val = line.partition(":")
                    key = key.strip()
                    val = val.strip().strip('"').strip("'")
                    if val.startswith("[") and val.endswith("]"):
                        val = val[1:-1].strip()
                    meta[key] = val
    
    # Extract H1
    h1 = re.search(r'^#\s+(.+)$', text, re.MULTILINE)
    if h1:
        meta['_title'] = h1.group(1).strip()
    
    return meta

def scan_people(brain: Path) -> list:
    people_dir = brain / "People"
    if not people_dir.is_dir():
        return []
    results = []
    for f in sorted(people_dir.glob("*.md")):
        m = parse_frontmatter(f)
        results.append({
            'file': f.stem,
            'name': m.get('_title', f.stem.replace('-', ' ').title()),
            'role': m.get('role', ''),
            'org': m.get('org', m.get('organization', '')),
            'last_contact': m.get('last_contact', m.get('updated', '')),
            'tags': m.get('tags', ''),
        })
    return results

def scan_projects(brain: Path) -> list:
    proj_dir = brain / "Projects"
    if not proj_dir.is_dir():
        return []
    results = []
    for d in sorted(proj_dir.iterdir()):
        if not d.is_dir():
            continue
        main_file = d / f"{d.name}.md"
        if not main_file.exists():
            # try any .md
            mds = list(d.glob("*.md"))
            main_file = mds[0] if mds else None
        if not main_file:
            continue
        m = parse_frontmatter(main_file)
        # Count sub-files
        sub_count = len(list(d.glob("*.md"))) - 1
        results.append({
            'file': d.name,
            'name': m.get('_title', d.name.replace('-', ' ').title()),
            'status': m.get('status', ''),
            'tags': m.get('tags', ''),
            'updated': m.get('updated', ''),
            'sub_notes': sub_count,
        })
    return results

def scan_areas(brain: Path) -> list:
    areas_dir = brain / "Areas"
    if not areas_dir.is_dir():
        return []
    results = []
    for f in sorted(areas_dir.glob("*.md")):
        m = parse_frontmatter(f)
        results.append({
            'file': f.stem,
            'name': m.get('_title', f.stem.replace('-', ' ').title()),
            'updated': m.get('updated', ''),
        })
    return results

def scan_notes(brain: Path) -> list:
    notes_dir = brain / "Notes"
    if not notes_dir.is_dir():
        return []
    results = []
    for f in sorted(notes_dir.glob("*.md")):
        m = parse_frontmatter(f)
        results.append({
            'file': f.stem,
            'name': m.get('_title', f.stem.replace('-', ' ').title()),
            'tags': m.get('tags', ''),
            'created': m.get('created', ''),
        })
    return results

def scan_tasks(brain: Path) -> dict:
    tasks_file = brain / "Tasks" / "tasks.md"
    if not tasks_file.exists():
        return {'focus': 0, 'next_up': 0, 'backlog': 0, 'done': 0}
    try:
        text = tasks_file.read_text(encoding="utf-8")
    except OSError:
        return {'focus': 0, 'next_up': 0, 'backlog': 0, 'done': 0}
    
    counts = {'focus': 0, 'next_up': 0, 'backlog': 0, 'done': 0}
    current_section = None
    for line in text.splitlines():
        lower = line.strip().lower()
        if lower.startswith('## focus'):
            current_section = 'focus'
        elif lower.startswith('## next up'):
            current_section = 'next_up'
        elif lower.startswith('## backlog'):
            current_section = 'backlog'
        elif lower.startswith('## done') or lower.startswith('## completed'):
            current_section = 'done'
        elif lower.startswith('## '):
            current_section = None
        elif current_section and re.match(r'^-\s*\[.\]', line.strip()):
            counts[current_section] += 1
    return counts

def scan_journal(brain: Path) -> dict:
    journal_dir = brain / "Journal"
    if not journal_dir.is_dir():
        return {'count': 0, 'latest': ''}
    entries = sorted(journal_dir.glob("*.md"))
    return {
        'count': len(entries),
        'latest': entries[-1].stem if entries else '',
    }

def format_full(brain: Path):
    lines = [f"# Knowledge Index\n**Brain:** `{brain}`\n"]
    
    # People
    people = scan_people(brain)
    if people:
        lines.append("## People\n| Name | Role/Org | Last Contact | Tags |")
        lines.append("|------|----------|--------------|------|")
        for p in people:
            org = f"{p['role']}" + (f" @ {p['org']}" if p['org'] else '')
            lines.append(f"| [[People/{p['file']}\\|{p['name']}]] | {org} | {p['last_contact']} | {p['tags']} |")
    else:
        lines.append("## People\n*None yet*")
    lines.append("")
    
    # Projects
    projects = scan_projects(brain)
    if projects:
        lines.append("## Projects\n| Name | Status | Updated | Sub-notes | Tags |")
        lines.append("|------|--------|---------|-----------|------|")
        for p in projects:
            lines.append(f"| [[Projects/{p['file']}/{p['file']}\\|{p['name']}]] | {p['status']} | {p['updated']} | {p['sub_notes']} | {p['tags']} |")
    else:
        lines.append("## Projects\n*None yet*")
    lines.append("")
    
    # Areas
    areas = scan_areas(brain)
    if areas:
        lines.append("## Areas\n| Name | Updated |")
        lines.append("|------|---------|")
        for a in areas:
            lines.append(f"| [[Areas/{a['file']}\\|{a['name']}]] | {a['updated']} |")
    else:
        lines.append("## Areas\n*None yet*")
    lines.append("")
    
    # Notes
    notes = scan_notes(brain)
    if notes:
        lines.append(f"## Notes ({len(notes)} total)\n| Name | Tags | Created |")
        lines.append("|------|------|---------|")
        for n in notes:
            lines.append(f"| [[Notes/{n['file']}\\|{n['name']}]] | {n['tags']} | {n['created']} |")
    else:
        lines.append("## Notes\n*None yet*")
    lines.append("")
    
    # Tasks
    tasks = scan_tasks(brain)
    lines.append(f"## Tasks\n- **Focus:** {tasks['focus']} items")
    lines.append(f"- **Next Up:** {tasks['next_up']} items")
    lines.append(f"- **Backlog:** {tasks['backlog']} items")
    lines.append(f"- **Done:** {tasks['done']} items")
    lines.append("")
    
    # Journal
    journal = scan_journal(brain)
    lines.append(f"## Journal\n- **Entries:** {journal['count']}")
    if journal['latest']:
        lines.append(f"- **Latest:** {journal['latest']}")
    
    return "\n".join(lines)

def format_compact(brain: Path):
    """One-line-per-category summary for hook injection."""
    people = [p['file'] for p in scan_people(brain)]
    projects = [p['file'] for p in scan_projects(brain)]
    areas = [a['file'] for a in scan_areas(brain)]
    notes = [n['file'] for n in scan_notes(brain)]
    tasks = scan_tasks(brain)
    journal = scan_journal(brain)
    
    lines = [
        f"BRAIN INDEX ({brain})",
        f"People: {', '.join(people) or 'none'}",
        f"Projects: {', '.join(projects) or 'none'}",
        f"Areas: {', '.join(areas) or 'none'}",
        f"Notes ({len(notes)}): {', '.join(notes[:20]) or 'none'}{'...' if len(notes) > 20 else ''}",
        f"Tasks: {tasks['focus']} focus, {tasks['next_up']} next, {tasks['backlog']} backlog",
        f"Journal: {journal['count']} entries, latest {journal['latest']}",
    ]
    return "\n".join(lines)

def main():
    brain = _brain_path()
    if not brain.is_dir():
        print(f"Brain not found at {brain}. Run /brain:init first.", file=sys.stderr)
        sys.exit(1)
    
    compact = "--compact" in sys.argv
    print(format_compact(brain) if compact else format_full(brain))

if __name__ == "__main__":
    main()

File: scripts/build-index.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const { getBrainRoot, parseFrontmatter } = require('./config');

function scanPeople(brain) {
  const dir = path.join(brain, 'People');
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir)
    .filter(f => f.endsWith('.md'))
    .sort()
    .map(f => {
      const m = parseFrontmatter(path.join(dir, f));
      const stem = f.slice(0, -3);
      return {
        file: stem,
        name: m._title || stem.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase()),
        role: m.role || '',
        org: m.org || m.organization || '',
        last_contact: m.last_contact || m.updated || '',
        tags: m.tags || '',
      };
    });
}

function scanProjects(brain) {
  const dir = path.join(brain, 'Projects');
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir)
    .filter(d => {
      try { return fs.statSync(path.join(dir, d)).isDirectory(); } catch { return false; }
    })
    .sort()
    .map(d => {
      const dirPath = path.join(dir, d);
      let mainFile = path.join(dirPath, `${d}.md`);
      if (!fs.existsSync(mainFile)) {
        const mds = fs.readdirSync(dirPath).filter(f => f.endsWith('.md'));
        mainFile = mds.length ? path.join(dirPath, mds[0]) : null;
      }
      if (!mainFile) return null;
      const m = parseFrontmatter(mainFile);
      const subCount = fs.readdirSync(dirPath).filter(f => f.endsWith('.md')).length - 1;
      return {
        file: d,
        name: m._title || d.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase()),
        status: m.status || '',
        tags: m.tags || '',
        updated: m.updated || '',
        sub_notes: Math.max(0, subCount),
      };
    })
    .filter(Boolean);
}

function scanAreas(brain) {
  const dir = path.join(brain, 'Areas');
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir)
    .filter(f => f.endsWith('.md'))
    .sort()
    .map(f => {
      const m = parseFrontmatter(path.join(dir, f));
      const stem = f.slice(0, -3);
      return {
        file: stem,
        name: m._title || stem.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase()),
        updated: m.updated || '',
      };
    });
}

function scanNotes(brain) {
  const dir = path.join(brain, 'Notes');
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir)
    .filter(f => f.endsWith('.md'))
    .sort()
    .map(f => {
      const m = parseFrontmatter(path.join(dir, f));
      const stem = f.slice(0, -3);
      return {
        file: stem,
        name: m._title || stem.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase()),
        tags: m.tags || '',
        created: m.created || '',
      };
    });
}

function scanTasks(brain) {
  const tasksFile = path.join(brain, 'Tasks', 'tasks.md');
  const counts = { focus: 0, next_up: 0, backlog: 0, done: 0 };
  let text;
  try { text = fs.readFileSync(tasksFile, 'utf-8'); } catch { return counts; }

  let current = null;
  for (const line of text.split('\n')) {
    const lower = line.trim().toLowerCase();
    if (lower.startsWith('## focus')) current = 'focus';
    else if (lower.startsWith('## next up')) current = 'next_up';
    else if (lower.startsWith('## backlog')) current = 'backlog';
    else if (lower.startsWith('## done') || lower.startsWith('## completed')) current = 'done';
    else if (lower.startsWith('## ')) current = null;
    else if (current && /^-\s*\[.\]/.test(line.trim())) counts[current]++;
  }
  return counts;
}

function scanJournal(brain) {
  const dir = path.join(brain, 'Journal');
  if (!fs.existsSync(dir)) return { count: 0, latest: '' };
  const entries = fs.readdirSync(dir).filter(f => f.endsWith('.md')).sort();
  return {
    count: entries.length,
    latest: entries.length ? entries[entries.length - 1].slice(0, -3) : '',
  };
}

function formatFull(brain) {
  const lines = [`# Knowledge Index\n**Brain:** \`${brain}\`\n`];

  const people = scanPeople(brain);
  if (people.length) {
    lines.push('## People\n| Name | Role/Org | Last Contact | Tags |');
    lines.push('|------|----------|--------------|------|');
    for (const p of people) {
      const org = p.role + (p.org ? ` @ ${p.org}` : '');
      lines.push(`| [[People/${p.file}\\|${p.name}]] | ${org} | ${p.last_contact} | ${p.tags} |`);
    }
  } else {
    lines.push('## People\n*None yet*');
  }
  lines.push('');

  const projects = scanProjects(brain);
  if (projects.length) {
    lines.push('## Projects\n| Name | Status | Updated | Sub-notes | Tags |');
    lines.push('|------|--------|---------|-----------|------|');
    for (const p of projects) {
      lines.push(`| [[Projects/${p.file}/${p.file}\\|${p.name}]] | ${p.status} | ${p.updated} | ${p.sub_notes} | ${p.tags} |`);
    }
  } else {
    lines.push('## Projects\n*None yet*');
  }
  lines.push('');

  const areas = scanAreas(brain);
  if (areas.length) {
    lines.push('## Areas\n| Name | Updated |');
    lines.push('|------|---------|');
    for (const a of areas) {
      lines.push(`| [[Areas/${a.file}\\|${a.name}]] | ${a.updated} |`);
    }
  } else {
    lines.push('## Areas\n*None yet*');
  }
  lines.push('');

  const notes = scanNotes(brain);
  if (notes.length) {
    lines.push(`## Notes (${notes.length} total)\n| Name | Tags | Created |`);
    lines.push('|------|------|---------|');
    for (const n of notes) {
      lines.push(`| [[Notes/${n.file}\\|${n.name}]] | ${n.tags} | ${n.created} |`);
    }
  } else {
    lines.push('## Notes\n*None yet*');
  }
  lines.push('');

  const tasks = scanTasks(brain);
  lines.push(`## Tasks\n- **Focus:** ${tasks.focus} items`);
  lines.push(`- **Next Up:** ${tasks.next_up} items`);
  lines.push(`- **Backlog:** ${tasks.backlog} items`);
  lines.push(`- **Done:** ${tasks.done} items`);
  lines.push('');

  const journal = scanJournal(brain);
  lines.push(`## Journal\n- **Entries:** ${journal.count}`);
  if (journal.latest) lines.push(`- **Latest:** ${journal.latest}`);

  return lines.join('\n');
}

function formatCompact(brain) {
  const people = scanPeople(brain).map(p => p.file);
  const projects = scanProjects(brain).map(p => p.file);
  const areas = scanAreas(brain).map(a => a.file);
  const notes = scanNotes(brain).map(n => n.file);
  const tasks = scanTasks(brain);
  const journal = scanJournal(brain);

  return [
    `BRAIN INDEX (${brain})`,
    `People: ${people.join(', ') || 'none'}`,
    `Projects: ${projects.join(', ') || 'none'}`,
    `Areas: ${areas.join(', ') || 'none'}`,
    `Notes (${notes.length}): ${notes.slice(0, 20).join(', ') || 'none'}${notes.length > 20 ? '...' : ''}`,
    `Tasks: ${tasks.focus} focus, ${tasks.next_up} next, ${tasks.backlog} backlog`,
    `Journal: ${journal.count} entries, latest ${journal.latest}`,
  ].join('\n');
}

module.exports = { scanPeople, scanProjects, scanAreas, scanNotes, scanTasks, scanJournal, formatFull, formatCompact };

if (require.main === module) {
  const brain = getBrainRoot();
  if (!fs.existsSync(brain)) {
    process.stderr.write(`Brain not found at ${brain}. Run /remember:init first.\n`);
    process.exit(1);
  }
  const compact = process.argv.includes('--compact');
  console.log(compact ? formatCompact(brain) : formatFull(brain));
}
